lib/web/desktop_playback.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/websocket"
+)
+
+const actionPlayPause = "play/pause"
+
+const (
+	playStatePlaying  = "playing"
+	playStatePaused   = "paused"
+	playStateFinished = "finished"
+)
+
+// playbackAction is a message passed from the playback client
+// to the server over the websocket connection in order to modify
+// the playback state.
+type playbackAction struct {
+	// Action is one of actionPlayPause | TODO: actionMove
+	// actionPlayPause toggles the playbackState.playState
+	Action string `json:"action"`
+}
+
+// playbackPlayer is a thread-safe struct for managing
+// the global state of the playback websocket connection.
+type playbackPlayer struct {
+	ws        *websocket.Conn
+	playState string // one of playStatePlaying | playStatePaused | playStateFinished
+	cond      *sync.Cond
+	mu        *sync.RWMutex
+	log       logrus.FieldLogger
+	closeOnce sync.Once
+	ctx       context.Context
+	cancel    context.CancelFunc
+	clt       *auth.Client
+	sID       string
+}
+
+func NewPlaybackPlayer(sID string, ws *websocket.Conn, clt *auth.Client, ctx context.Context, log logrus.FieldLogger) *playbackPlayer {
+	ws.PayloadType = websocket.BinaryFrame
+	var mu sync.RWMutex
+	cond := sync.NewCond(&mu)
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &playbackPlayer{
+		ws:        ws,
+		playState: playStatePlaying,
+		cond:      cond,
+		mu:        &mu,
+		log:       log,
+		ctx:       ctx,
+		cancel:    cancel,
+		clt:       clt,
+		sID:       sID,
+	}
+}
+
+// hangWhilePaused hangs while pp.playState = playStatePaused. When hangWhilePaused is called in one goroutine,
+// it will only be triggered to check its condition again by another goroutine calling pp.TogglePlaying() or pp.Close()
+// (because those functions call pp.cond.Broadcast()).
+func (pp *playbackPlayer) hangWhilePaused() {
```

Comment:
Let's call this `waitForPlay` or something similar. Hang has a negative connotation to it.